File: scripts/vref_sweep/plot_vref_sweep.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import csv
import math
import re
import sys
import zipfile
from pathlib import Path
from typing import Any
from xml.etree import ElementTree as ET


NS_MAIN = "{http://schemas.openxmlformats.org/spreadsheetml/2006/main}"


class VrefPlotError(RuntimeError):
    pass


def parse_float(v: Any) -> float | None:
    if v is None:
        return None
    s = str(v).strip()
    if not s:
        return None
    try:
        x = float(s)
    except ValueError:
        return None
    if not math.isfinite(x):
        return None
    return x


def col_index_from_cell_ref(ref: str) -> int:
    letters = "".join(ch for ch in str(ref) if ch.isalpha()).upper()
    if not letters:
        return 0
    idx = 0
    for ch in letters:
        idx = idx * 26 + (ord(ch) - ord("A") + 1)
    return idx - 1


def read_shared_strings(zf: zipfile.ZipFile) -> list[str]:
    try:
        raw = zf.read("xl/sharedStrings.xml")
    except KeyError:
        return []
    root = ET.fromstring(raw)
    strings: list[str] = []
    for si in root.findall(f"{NS_MAIN}si"):
        parts: list[str] = []
        for t in si.iter(f"{NS_MAIN}t"):
            parts.append(t.text or "")
        strings.append("".join(parts))
    return strings


def read_xlsx_rows(path: Path) -> list[list[Any]]:
    with zipfile.ZipFile(path) as zf:
        shared = read_shared_strings(zf)
        sheet = ET.fromstring(zf.read("xl/worksheets/sheet1.xml"))

    rows: list[list[Any]] = []
    for row_el in sheet.iter(f"{NS_MAIN}row"):
        vals: dict[int, Any] = {}
        for cell in row_el.findall(f"{NS_MAIN}c"):
            ref = cell.attrib.get("r", "")
            col = col_index_from_cell_ref(ref)
            ctype = cell.attrib.get("t", "")
            value: Any = None
            if ctype == "inlineStr":
                texts = [t.text or "" for t in cell.iter(f"{NS_MAIN}t")]
                value = "".join(texts)
            else:
                v_el = cell.find(f"{NS_MAIN}v")
                raw = v_el.text if v_el is not None else None
                if ctype == "s" and raw is not None:
                    try:
                        value = shared[int(raw)]
                    except Exception:
                        value = raw
                elif raw is not None:
                    num = parse_float(raw)
                    value = num if num is not None else raw
            vals[col] = value
        if vals:
            max_col = max(vals)
            rows.append([vals.get(i) for i in range(max_col + 1)])
    return rows


def normalize_header(s: Any) -> str:
    return re.sub(r"[^a-z0-9]+", "", str(s or "").strip().lower())


def find_column(headers: list[Any], candidates: set[str]) -> int | None:
    norm = [normalize_header(h) for h in headers]
    for idx, h in enumerate(norm):
        if h in candidates:
            return idx
    return None


def read_curve_from_xlsx(path: Path, *, complete_only: bool, min_fail_cw: int) -> list[tuple[float, float]]:
    rows = read_xlsx_rows(path)
    if not rows:
        return []
    headers = rows[0]
    raw_idx = find_column(headers, {"rawber", "rber"})
    fer_idx = find_column(headers, {"ldpcfer", "fer"})
    fail_idx = find_column(headers, {"failcw", "fail", "failcount"})
    complete_idx = find_column(headers, {"iscomplete", "complete"})
    if raw_idx is None or fer_idx is None:
        raise VrefPlotError(f"Cannot find RAW_BER/LDPC_FER columns in {path}")
    if min_fail_cw > 0 and fail_idx is None:
        raise VrefPlotError(f"--min-fail-cw requires a FAIL_CW column in {path}")

    pts: list[tuple[float, float]] = []
    for row in rows[1:]:
        if complete_only and complete_idx is not None:
            if complete_idx >= len(row) or parse_float(row[complete_idx]) != 1.0:
                continue
        if min_fail_cw > 0:
            fail_cw = parse_float(row[fail_idx] if fail_idx is not None and fail_idx < len(row) else None)
            if fail_cw is None or fail_cw < min_fail_cw:
                continue
        raw = parse_float(row[raw_idx] if raw_idx < len(row) else None)
        fer = parse_float(row[fer_idx] if fer_idx < len(row) else None)
        if raw is None or fer is None or raw <= 0.0 or fer <= 0.0:
            continue
        pts.append((raw, fer))
    pts.sort(key=lambda x: x[0])
    return pts


def load_metadata(path: Path) -> dict[str, dict[str, str]]:
    if not path.exists():
        return {}
    out: dict[str, dict[str, str]] = {}
    with path.open("r", encoding="utf-8", newline="") as fh:
        reader = csv.DictReader(fh)
        for row in reader:
            case_name = str(row.get("case_name", "")).strip()
            if case_name:
                out[case_name] = {str(k): "" if v is None else str(v) for k, v in row.items()}
            progress = str(row.get("progress_xlsx", "")).strip()
            if progress:
                out[str(Path(progress).resolve())] = {str(k): "" if v is None else str(v) for k, v in row.items()}
    return out


def case_name_from_xlsx(path: Path) -> str:
    stem = path.stem
    if stem.endswith("_progress"):
        return stem[: -len("_progress")]
    return stem


def make_vref_label(value: str, fallback: str) -> str:
    vref = str(value or "").strip() or fallback
    if vref.lower().startswith("vref="):
        return vref
    return f"VREF={vref}"


def label_for(path: Path, metadata: dict[str, dict[str, str]]) -> str:
    row = metadata.get(str(path.resolve()))
    if row is None:
        row = metadata.get(case_name_from_xlsx(path))
    fallback = case_name_from_xlsx(path)
    if row is None:
        return make_vref_label("", fallback)
    vref = row.get("vref_label") or row.get("vref_values") or ""
    return make_vref_label(vref, fallback)


def discover_xlsx(input_root: Path) -> list[Path]:
    return sorted(p.resolve() for p in input_root.rglob("*_progress.xlsx") if p.is_file())


def plot_single(path: Path, pts: list[tuple[float, float]], label: str, out_path: Path) -> None:
    import matplotlib.pyplot as plt  # type: ignore

    out_path.parent.mkdir(parents=True, exist_ok=True)
    xs = [p[0] for p in pts]
    ys = [p[1] for p in pts]
    plt.figure(figsize=(8, 5.5))
    plt.semilogy(xs, ys, marker="o", linewidth=1.4, markersize=4, label=label)
    plt.xlabel("RBER")
    plt.ylabel("FER")
    plt.grid(True, which="both", linestyle="--", linewidth=0.5)
    plt.legend(fontsize=7)
    plt.tight_layout()
    plt.savefig(out_path, dpi=180)
    plt.close()


def is_baseline_curve(path: Path, label: str) -> bool:
    case = case_name_from_xlsx(path).lower()
    label_lower = label.lower()
    return case == "baseline" or label_lower.startswith("baseline")


def plot_overlay(curves: list[tuple[Path, str, list[tuple[float, float]]]], out_path: Path) -> None:
    import matplotlib.pyplot as plt  # type: ignore

    out_path.parent.mkdir(parents=True, exist_ok=True)
    plt.figure(figsize=(11, 7))
    try:
        cmap = plt.get_cmap("turbo", max(len(curves), 1))
    except ValueError:
        cmap = plt.get_cmap("hsv", max(len(curves), 1))

    for idx, (path, label, pts) in enumerate(curves):
        xs = [p[0] for p in pts]
        ys = [p[1] for p in pts]
        linestyle = "--" if is_baseline_curve(path, label) else "-"
        plt.semilogy(xs, ys, linestyle=linestyle, color=cmap(idx), marker="o", linewidth=1.1, markersize=3, label=label)

    plt.xlabel("RBER")
    plt.ylabel("FER")
    plt.grid(True, which="both", linestyle="--", linewidth=0.5)
    plt.legend(fontsize=6, loc="center left", bbox_to_anchor=(1.02, 0.5), borderaxespad=0.0)
    plt.tight_layout(rect=(0, 0, 0.72, 1))
    plt.savefig(out_path, dpi=180)
    plt.close()


def parse_args(argv: list[str]) -> argparse.Namespace:
    ap = argparse.ArgumentParser(description="Plot VREF sweep RBER-FER curves from AutoFER progress xlsx")
    ap.add_argument("--input-root", default="output/vref_sweep", help="Root containing AutoFER *_progress.xlsx files")
    ap.add_argument("--metadata", default="", help="cases.csv generated by submit_vref_sweep.py")
    ap.add_argument("--out-dir", default="", help="Plot output directory (default: <input-root>/plots)")
    ap.add_argument("--complete-only", action="store_true", help="Only plot rows with IsComplete=1")
    ap.add_argument("--min-fail-cw", type=int, default=0, help="Only plot rows with FAIL_CW >= this value (default: 0)")
    return ap.parse_args(argv)


def main(argv: list[str]) -> int:
    args = parse_args(argv)
    input_root = Path(args.input_root).expanduser().resolve()
    if not input_root.exists():
        raise VrefPlotError(f"input root not found: {input_root}")
    metadata_path = Path(args.metadata).expanduser().resolve() if str(args.metadata).strip() else input_root / "cases.csv"
    metadata = load_metadata(metadata_path)
    out_dir = Path(args.out_dir).expanduser().resolve() if str(args.out_dir).strip() else input_root / "plots"

    xlsx_files = discover_xlsx(input_root)
    if not xlsx_files:
        raise VrefPlotError(f"No *_progress.xlsx files found under {input_root}")

    curves: list[tuple[Path, str, list[tuple[float, float]]]] = []
    for xlsx in xlsx_files:
        pts = read_curve_from_xlsx(
            xlsx,
            complete_only=bool(args.complete_only),
            min_fail_cw=max(0, int(args.min_fail_cw)),
        )
        if not pts:
            continue
        label = label_for(xlsx, metadata)
        curves.append((xlsx, label, pts))
        plot_single(xlsx, pts, label, out_dir / "single" / f"{case_name_from_xlsx(xlsx)}.png")

    if not curves:
        raise VrefPlotError("No plottable RBER/FER points found")
    plot_overlay(curves, out_dir / "all_vref_overlay.png")
    print(f"plotted {len(curves)} curve(s) to {out_dir}")
    return 0


if __name__ == "__main__":
    try:
        raise SystemExit(main(sys.argv[1:]))
    except VrefPlotError as exc:
        print(f"[ERROR] {exc}", file=sys.stderr)
        raise SystemExit(1)

File: scripts/vref_sweep/plot_vref_sweep.m
function plot_vref_sweep(varargin)
%PLOT_VREF_SWEEP Plot VREF sweep RBER-FER curves from AutoFER progress xlsx.
%
% Command-style usage:
%   plot_vref_sweep('--input-root', 'output/vref_sweep', ...
%                   '--out-dir', 'output/vref_sweep/plots_matlab', ...
%                   '--min-fail-cw', 10)
%
% Name-value usage:
%   plot_vref_sweep('input_root', 'output/vref_sweep', ...
%                   'min_fail_cw', 10)

opts = parse_args(varargin{:});

input_root = char(opts.input_root);
if ~isfolder(input_root)
    error('plot_vref_sweep:InputRootNotFound', 'input root not found: %s', input_root);
end

if strlength(opts.metadata) > 0
    metadata_path = char(opts.metadata);
else
    metadata_path = fullfile(input_root, 'cases.csv');
end

if strlength(opts.out_dir) > 0
    out_dir = char(opts.out_dir);
else
    out_dir = fullfile(input_root, 'plots_matlab');
end

metadata = load_metadata(metadata_path);
xlsx_files = discover_xlsx(input_root);
if isempty(xlsx_files)
    error('plot_vref_sweep:NoXlsx', 'No *_progress.xlsx files found under %s', input_root);
end

curves = struct('path', {}, 'label', {}, 'x', {}, 'y', {}, 'case_name', {});
for i = 1:numel(xlsx_files)
    xlsx_path = xlsx_files{i};
    [x, y] = read_curve_from_xlsx(xlsx_path, opts.complete_only, opts.min_fail_cw);
    if isempty(x)
        continue;
    end

    cname = case_name_from_xlsx(xlsx_path);
    lbl = label_for(xlsx_path, cname, metadata);
    curves(end + 1).path = xlsx_path; %#ok<AGROW>
    curves(end).label = lbl;
    curves(end).x = x;
    curves(end).y = y;
    curves(end).case_name = cname;

end

if isempty(curves)
    error('plot_vref_sweep:NoPoints', 'No plottable RBER/FER points found');
end

ensure_dir(out_dir);
fig = figure('Name', 'all_vref_overlay', 'NumberTitle', 'off', 'Color', 'w');
hold on;

colors = make_colors(numel(curves));
for i = 1:numel(curves)
    if is_baseline_curve(curves(i))
        line_style = '--o';
    else
        line_style = '-o';
    end
    semilogy(curves(i).x, curves(i).y, line_style, ...
        'Color', colors(i, :), 'LineWidth', 1.1, 'MarkerSize', 4, ...
        'DisplayName', curves(i).label);
end
set(gca, 'YScale', 'log');
hold off;
xlabel('RBER');
ylabel('FER');
grid on;
legend('show', 'Interpreter', 'none', 'Location', 'eastoutside');
title('VREF Sweep RBER-FER', 'Interpreter', 'none');
save_figure(fig, fullfile(out_dir, 'all_vref_overlay'));

fprintf('plotted %d curve(s) to %s\n', numel(curves), out_dir);
end


function tf = is_baseline_curve(curve)
cname = lower(string(curve.case_name));
lbl = lower(string(curve.label));
tf = cname == "baseline" || startsWith(lbl, "baseline");
end


function colors = make_colors(n)
if n <= 0
    colors = zeros(0, 3);
elseif exist('turbo', 'builtin') || exist('turbo', 'file')
    colors = turbo(n);
else
    colors = hsv(n);
end
end


function opts = parse_args(varargin)
opts = struct();
opts.input_root = "output/vref_sweep";
opts.metadata = "";
opts.out_dir = "";
opts.complete_only = false;
opts.min_fail_cw = 0;

args = varargin;
if numel(args) == 1 && iscell(args{1})
    args = args{1};
end

i = 1;
while i <= numel(args)
    key = string(args{i});
    key = strip(key);
    if startsWith(key, "--")
        key = extractAfter(key, 2);
    end
    key = strrep(key, "-", "_");

    switch lower(char(key))
        case 'input_root'
            [opts.input_root, i] = read_value(args, i);
        case 'metadata'
            [opts.metadata, i] = read_value(args, i);
        case 'out_dir'
            [opts.out_dir, i] = read_value(args, i);
        case 'complete_only'
            if i == numel(args) || startsWith(string(args{i + 1}), "--")
                opts.complete_only = true;
                i = i + 1;
            else
                [v, i] = read_value(args, i);
                opts.complete_only = parse_bool(v);
            end
        case 'min_fail_cw'
            [v, i] = read_value(args, i);
            opts.min_fail_cw = max(0, round(str2double(string(v))));
            if isnan(opts.min_fail_cw)
                error('plot_vref_sweep:BadMinFailCw', 'min_fail_cw must be numeric');
            end
        otherwise
            error('plot_vref_sweep:BadArg', 'Unknown argument: %s', args{i});
    end
end
end


function [value, next_i] = read_value(args, i)
if i >= numel(args)
    error('plot_vref_sweep:MissingValue', 'Missing value for argument %s', args{i});
end
value = string(args{i + 1});
next_i = i + 2;
end


function v = parse_bool(x)
s = lower(strip(string(x)));
v = any(s == ["true", "1", "yes", "y", "on"]);
end


function files = discover_xlsx(input_root)
d = dir(fullfile(input_root, '**', '*_progress.xlsx'));
files = cell(numel(d), 1);
for i = 1:numel(d)
    files{i} = fullfile(d(i).folder, d(i).name);
end
files = sort(files);
end


function metadata = load_metadata(metadata_path)
metadata = struct('case_name', {}, 'progress_xlsx', {}, 'vref_label', {}, 'vref_values', {});
if ~isfile(metadata_path)
    return;
end

T = readtable(metadata_path, 'FileType', 'text', 'Delimiter', ',', 'VariableNamingRule', 'preserve');
names = T.Properties.VariableNames;
for i = 1:height(T)
    rec = struct();
    rec.case_name = string(get_table_value(T, names, i, 'case_name'));
    rec.progress_xlsx = string(get_table_value(T, names, i, 'progress_xlsx'));
    rec.vref_label = string(get_table_value(T, names, i, 'vref_label'));
    rec.vref_values = string(get_table_value(T, names, i, 'vref_values'));
    metadata(end + 1) = rec; %#ok<AGROW>
end
end


function value = get_table_value(T, names, row_idx, col_name)
idx = find(strcmp(names, col_name), 1);
if isempty(idx)
    value = "";
    return;
end
col = T.(names{idx});
value = col(row_idx);
if iscell(value)
    value = value{1};
end
if ismissing(value)
    value = "";
end
end


function [x, y] = read_curve_from_xlsx(xlsx_path, complete_only, min_fail_cw)
T = readtable(xlsx_path, 'FileType', 'spreadsheet', 'VariableNamingRule', 'preserve');
headers = T.Properties.VariableNames;

raw_idx = find_column(headers, ["rawber", "rber"]);
fer_idx = find_column(headers, ["ldpcfer", "fer"]);
fail_idx = find_column(headers, ["failcw", "fail", "failcount"]);
complete_idx = find_column(headers, ["iscomplete", "complete"]);

if isempty(raw_idx) || isempty(fer_idx)
    error('plot_vref_sweep:MissingColumns', 'Cannot find RAW_BER/LDPC_FER columns in %s', xlsx_path);
end
if min_fail_cw > 0 && isempty(fail_idx)
    error('plot_vref_sweep:MissingFailCw', '--min-fail-cw requires a FAIL_CW column in %s', xlsx_path);
end

raw = to_double(T.(headers{raw_idx}));
fer = to_double(T.(headers{fer_idx}));
keep = isfinite(raw) & isfinite(fer) & raw > 0 & fer > 0;

if complete_only && ~isempty(complete_idx)
    complete = to_double(T.(headers{complete_idx}));
    keep = keep & complete == 1;
end

if min_fail_cw > 0
    fail_cw = to_double(T.(headers{fail_idx}));
    keep = keep & isfinite(fail_cw) & fail_cw >= min_fail_cw;
end

x = raw(keep);
y = fer(keep);
[x, order] = sort(x);
y = y(order);
end


function idx = find_column(headers, candidates)
idx = [];
for i = 1:numel(headers)
    h = normalize_header(headers{i});
    if any(h == candidates)
        idx = i;
        return;
    end
end
end


function h = normalize_header(x)
h = lower(regexprep(string(x), '[^a-zA-Z0-9]+', ''));
end


function out = to_double(v)
if isnumeric(v) || islogical(v)
    out = double(v);
    return;
end
if iscell(v)
    out = nan(numel(v), 1);
    for i = 1:numel(v)
        out(i) = str2double(string(v{i}));
    end
    return;
end
out = str2double(string(v));
end


function cname = case_name_from_xlsx(xlsx_path)
[~, stem, ~] = fileparts(xlsx_path);
suffix = '_progress';
if endsWith(stem, suffix)
    cname = extractBefore(string(stem), strlength(stem) - strlength(suffix) + 1);
else
    cname = string(stem);
end
cname = char(cname);
end


function lbl = label_for(xlsx_path, cname, metadata)
abs_xlsx = string(abs_path(xlsx_path));
for i = 1:numel(metadata)
    if metadata(i).progress_xlsx ~= ""
        if string(abs_path(metadata(i).progress_xlsx)) == abs_xlsx
            lbl = make_label(cname, metadata(i));
            return;
        end
    end
end
for i = 1:numel(metadata)
    if metadata(i).case_name == string(cname)
        lbl = make_label(cname, metadata(i));
        return;
    end
end
lbl = make_vref_label("", cname);
end


function lbl = make_label(cname, rec)
vref = rec.vref_label;
if vref == ""
    vref = rec.vref_values;
end
lbl = make_vref_label(vref, cname);
end


function lbl = make_vref_label(value, fallback)
vref = strip(string(value));
if vref == ""
    vref = string(fallback);
end
if startsWith(lower(vref), "vref=")
    lbl = char(vref);
else
    lbl = char("VREF=" + vref);
end
end


function p = abs_path(path_in)
p = char(path_in);
if exist('javaObject', 'builtin') || exist('javaObject', 'file')
    try
        p = char(java.io.File(p).getCanonicalPath());
        return;
    catch
    end
end
if ~isfolder(p) && ~isfile(p)
    [folder, name, ext] = fileparts(p);
    if isempty(folder)
        folder = pwd;
    end
    p = fullfile(folder, [name ext]);
end
end


function ensure_dir(path_in)
if ~isfolder(path_in)
    mkdir(path_in);
end
end


function save_figure(fig, base_path)
savefig(fig, [base_path '.fig']);
try
    exportgraphics(fig, [base_path '.png'], 'Resolution', 200);
catch
    print(fig, [base_path '.png'], '-dpng', '-r200');
end
end
